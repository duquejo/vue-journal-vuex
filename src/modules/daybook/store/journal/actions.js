import journalApi from "../../../../api/journalApi";

export const loadEntries = async ({ commit }) => {
  const { data } = await journalApi.get('/entries.json');

  if( ! data ) {
    commit('setEntries', []);
    return;
  }
  const entries = [];
  for (let id of Object.keys(data)) {
    entries.push({
      id,
      ...data[id],
    });
  }
  commit('setEntries', entries);
};
export const updateEntry = async ({ commit }, entry ) => {
  const { id, date, picture, text  } = entry;

  const dataToSave = { date, picture, text };

  await journalApi.put(`/entries/${id}.json`, dataToSave );

  dataToSave.id = id;

  commit('updateEntry', { ...dataToSave });
};
export const createEntry = async ({ commit }, entry ) => {
  const { date, picture, text } = entry;

  const dataToSave = { date, picture, text };

  const { data } = await journalApi.post('/entries.json', dataToSave );
  commit('addEntry', { id: data.name, ...dataToSave });
  return data.name;
};
export const deleteEntry = async ({ commit }, id ) => {
  await journalApi.delete(`/entries/${id}.json` );
  commit('deleteEntry', id );
  return id;
};