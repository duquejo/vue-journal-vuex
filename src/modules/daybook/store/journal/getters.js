export const getEntriesByTerm = (/*state*/) => {
};
export const getEntryById = (/*state*/) => {
};