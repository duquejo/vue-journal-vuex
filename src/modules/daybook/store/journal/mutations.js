export const setEntries = (/*state*/) => {
};
export const setLoading = (/*state*/) => {
};
export const updateEntry = (/*state*/) => {
};
export const addEntry = (/*state*/) => {
};