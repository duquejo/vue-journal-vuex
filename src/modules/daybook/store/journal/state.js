export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(), // 12315646734
      date: new Date().toDateString(), // Sat 23, July
      text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Soluta recusandae iusto harum sint maxime laborum.',
      picture: null,
    },
    {
      id: new Date().getTime() + 1000, // 12315646734
      date: new Date().toDateString(), // Sat 23, July
      text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Molestias, blanditiis.',
      picture: null,
    },
    {
      id: new Date().getTime() + 2000, // 12315646734
      date: new Date().toDateString(), // Sat 23, July
      text: 'Lorem ipsum dolor sit amet.',
      picture: null,
    },
  ]
});