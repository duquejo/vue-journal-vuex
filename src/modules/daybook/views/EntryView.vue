<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-3">{{ month }}</span>
        <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
      </div>
      <div>

        <input
          type="file"
          @change="onSelectedImage"
          ref="imageSelector"
          v-show="false"
          accept="image/png, image/jpeg"
        />

        <button v-if="entry.id" 
          @click="onDeleteEntry" 
          class="btn btn-danger mx-2">
          Erase
          <i class="fa fa-trash-alt"></i>
        </button>
        <button @click="onSelectImage" class="btn btn-primary">
          Upload photo
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>
    <hr>
    <div class="d-flex flex-column px-3 h-75">
      <textarea
        v-model="entry.text"
        placeholder="What happened today?"></textarea>
    </div>
    <img
      v-if="entry.picture && ! localImage"
      :src="entry.picture" 
      alt="entry picture"
      class="img-thumbnail"
    />
    <img
      v-if="localImage"
      :src="localImage"
      alt="entry picture"
      class="img-thumbnail"
    />
  </template>

  <FabComponent
    icon="fa-save"
    @on:click="saveEntry"
  />
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapActions, mapGetters } from 'vuex';
import Swal from 'sweetalert2';
import getDayMonthYear from '../helpers/getDayMonthYear';
import uploadImage from '../helpers/uploadImage';

export default {
  name: 'EntryView',
  props: {
    id: {
      type: String,
      required: true,
    }
  },
  data() {
    return {
      entry: null,
      localImage: null,
      file: null,
    };
  },
  methods: {
    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
    loadEntry() {
      let entry;
      if( this.id === 'new' ) {
        entry = {
          text: '',
          date: new Date().getTime(),
        };
      } else {
        entry = this.getEntryById(this.id);
        if( ! entry ) {
          return this.$router.push({ name: 'no-entry' });
        }
      }
      this.entry = entry;
    },
    async saveEntry() {

      Swal.fire({
        title: 'Wait a sec...',
        allowOutsideClick: false,
      });
      Swal.showLoading();

      const picture = await uploadImage( this.file );
      this.entry.picture = picture;

      if( this.entry.id ) {
        /**
         * Update
         */
        await this.updateEntry(this.entry);
      } else {
        /**
         * New entry
         */
       const id = await this.createEntry(this.entry);
       this.$router.push({ name: 'entry', params: { id } });
      }

      this.file = null;
      this.localImage = null;
      Swal.fire('Saving', 'Entry registered successfully', 'success');
    },
    async onDeleteEntry() {

      const { isConfirmed } = await Swal.fire({
        title: 'Are you sure?',
        text: 'Once deleted, it cannot be recovered.',
        showDenyButton: true,
        confirmButtonText: 'Yes, I\'m sure',
      });

      if( isConfirmed ) {
        await Swal.fire({
          title: 'Wait a sec...',
          allowOutsideClick: false,
        });
        Swal.showLoading();
        await this.deleteEntry(this.entry.id);
        this.$router.push({ name: 'no-entry' });
        Swal.fire('Erased', '', 'success');
      }

    },
    onSelectedImage( event ) {
      const file = event.target.files[0];
      if( !file ) {
        this.localImage = null;
        this.file = null;
        return;
      }
      this.file =  file;
      const fr = new FileReader();
      fr.onload = () => this.localImage = fr.result;
      fr.readAsDataURL(file);
    },
    onSelectImage() {
      this.$refs.imageSelector.click();
    },
  },
  computed: {
    ...mapGetters('journal', ['getEntryById']),
    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    yearDay() {
      const { yearDay } = getDayMonthYear(this.entry.date);
      return yearDay;
    },
  },
  created() {
    this.localImage = null;
    this.file = null;
    this.loadEntry();
  },
  watch: {
    id() {
      this.localImage = null;
      this.file = null;
      this.loadEntry(); // Entry watcher.
    },
  },
  components: {
    FabComponent: defineAsyncComponent(() => import('../components/FabComponent')),
  },
}
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;
  &:focus{
    outline: none;
  }
}
img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0 5px 10px rgba($color: #000000, $alpha: 0.2);
  pointer-events: none;
}
</style>