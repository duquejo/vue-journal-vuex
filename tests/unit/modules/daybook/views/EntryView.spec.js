import { createStore } from 'vuex';
import journal from '@/modules/daybook/store/journal';
import { shallowMount } from '@vue/test-utils';
import EntryView from '@/modules/daybook/views/EntryView';
import { journalState } from '../../../mocks/test-journal-state';
import Swal from 'sweetalert2';

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn(),
}));

describe('EntryView.vue', () => {

  const routerMock = {
    push: jest.fn(),
  };

  const createVuexStore = (initialState) => createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...initialState },
      },
    },
  });

  const store = createVuexStore(journalState);
  store.dispatch = jest.fn(); // Mocking inner functions. (because it could trigger API calls.)

  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryView,{
      props: {
        id: journalState.entries[0].id,
      },
      global: {
        mocks: {
          $router: routerMock,
        },
        plugins: [ store ],
      },
    });
  });

  it('should redirect the user when the ID doesn\'t exists.', () => {
    shallowMount(EntryView, {
      props: {
        id: 'This ID doesn\'t exists!',
      },
      global: {
        mocks: {
          $router: routerMock,
        },
        plugins: [ store ],
      },
    });
    expect(routerMock.push).toHaveBeenCalled();
    expect(routerMock.push).toHaveBeenCalledTimes(1);
    expect(routerMock.push).toHaveBeenCalledWith({ name: 'no-entry' });
  });


  it('should show the entry as well.', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(routerMock.push).not.toHaveBeenCalled();
  });

  it('should remove an entry and it should redirect.', (done) => {
    Swal.fire.mockResolvedValueOnce({ isConfirmed: true });

    wrapper.find('.btn-danger').trigger('click');

    expect(Swal.fire).toHaveBeenCalled();
    expect(Swal.fire).toHaveBeenCalledWith({
      title: 'Are you sure?',
      text: 'Once deleted, it cannot be recovered.',
      showDenyButton: true,
      confirmButtonText: 'Yes, I\'m sure',
    });

    setTimeout(() => {
      expect(store.dispatch).toHaveBeenCalled();
      expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', journalState.entries[0].id);
      expect(routerMock.push).toHaveBeenCalled();
      expect(routerMock.push).toHaveBeenCalledWith({ name: 'no-entry' });
      done();
    }, 1);

  });
});