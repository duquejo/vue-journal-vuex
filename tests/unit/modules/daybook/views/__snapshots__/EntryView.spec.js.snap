// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView.vue should show the entry as well. 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">9</span><span class="mx-1 fs-3">September</span><span class="mx-2 fs-4 fw-light">2023, Saturday</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Erase <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload photo <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What happened today?"></textarea></div>
<img src="https://abc_xyz.jpg" alt="entry picture" class="img-thumbnail">
<!--v-if-->
<fab-component-stub icon="fa-save"></fab-component-stub>
`;
